Cam wp7 task3

Mikolaj-A-Kowalski opened the pull request:
This pull request addresses issue #1470 and should fulfil requirements of task 3. 

Major changes: 

 * Adds option to ``Element.expand`` to enrich an arbitrary element composed of two isotopes 
 * ``enrichment`` argument specifies the enrichment in %; ``enrichment_target`` specifies target isotope e.g. 'Li7' and ``enrichment_type`` is either 'ao' or 'wo' to tell is the enrichment is in terms of atoms or mass. I am a bit warried about the large number of optional arguments now present in ``Element.expand`` and ``Material.add_element``, but I had no idea how to reduce them  
* I added new unit_test (``test_element``) that enforces the behaviour enrichment in  ``Element.expand`` 
* Extra cases were also added to ``test_material`` unit test to cover correct enrichment for Li and H as well as test for exception with incorrect settings for U and He

Commit:
Added internal infrastructure for dlopen based shared object sources

Commit:
Added internal infrastructure for dlopen based shared object sources

Commit:
Added internal infrastructure for dlopen based shared object sources

Commit:
Added internal infrastructure for dlopen based shared object sources

Commit:
Added internal infrastructure for dlopen based shared object sources

Commit:
Added internal infrastructure for dlopen based shared object sources

Commit:
Added missing fixed source subroutine

Commit:
Added missing fixed source subroutine

Commit:
Added missing fixed source subroutine

Commit:
Added missing fixed source subroutine

Commit:
Added missing fixed source subroutine

Commit:
Added missing fixed source subroutine

Commit:
Added example on how to use xml based custom source

Commit:
Added example on how to use xml based custom source

Commit:
Added python bindngs for library based source

Commit:
Added python bindngs for library based source

Commit:
Added python bindngs for library based source

Commit:
Added python bindngs for library based source

Commit:
Added python bindngs for library based source

Commit:
Added python bindngs for library based source

makeclean committed:
Fix the python part

makeclean committed:
Fix the python part

makeclean committed:
Fix the python part

makeclean committed:
Fix the python part

makeclean committed:
Fix the python part

makeclean committed:
Fix the python part

makeclean committed:
Now writes a Makefile for users to build a source

makeclean committed:
Now writes a Makefile for users to build a source

makeclean committed:
missing space

makeclean committed:
missing space

makeclean committed:
Adding unit test for dlopen source

makeclean committed:
Adding unit test for dlopen source

makeclean committed:
make a cmakefile instead of a makefile for cross platform compatibility

makeclean committed:
make a cmakefile instead of a makefile for cross platform compatibility

makeclean committed:
Added regression tests for dlopen source

makeclean committed:
Added regression tests for dlopen source

makeclean committed:
Stylistic changes

makeclean committed:
Stylistic changes

makeclean committed:
added unit test for dlopen source, and changed library variable name

makeclean committed:
added unit test for dlopen source, and changed library variable name

makeclean committed:
Added supporting documentation

makeclean committed:
Added supporting documentation

makeclean committed:
Wrapped the first instance trying to use dlopen in a posix safety ifdef

makeclean committed:
Wrapped the first instance trying to use dlopen in a posix safety ifdef

makeclean committed:
made the building of the cmake more sensible

makeclean committed:
made the building of the cmake more sensible

makeclean committed:
try adding an init py file

makeclean committed:
try adding an init py file

makeclean committed:
cxx standards for the source compile

makeclean committed:
cxx standards for the source compile

Mikolaj-A-Kowalski committed:
Change interface of Element.expand function

-> Adds new optional argument enrichment_target to Element.expand
-> expand uses old Uranium Enrichment method if enrichment is given
without enrichment target
-> Adds a comment to openmc.data to clarify that NATURAL_ABUNDANCE
contains atomic fractions

Mikolaj-A-Kowalski committed:
Initial enrichment option in Element.expand

-> Implementation supports only wt% enrichment
-> Multiple unnecessary conversions from at% to wt% can happen
-> Can produce -ve fractions if enrichment > 100

Mikolaj-A-Kowalski committed:
Fills gaps in docstring of Element

eepeterson committed:
refactoring surface classes

eepeterson committed:
working on quadric refactoring

eepeterson committed:
finished skeleton framework for refactoring surfaces

eepeterson committed:
working out some _get_base_coeff methods

eepeterson committed:
fixed typos in translate and __init__

eepeterson committed:
still working through coefficient transformations, done with (X,Y,Z)Cylinders and Cones, Planes, Quadric

eepeterson committed:
cleaning up and making code more self-consistent

eepeterson committed:
fixed issue related to finding all surface subclasses

eepeterson committed:
fixed bug in quadric translate function

eepeterson committed:
refactoring bounding_box and cleaning up constructors

eepeterson committed:
changed subclass finding methods to private

eepeterson committed:
removed empty dictionary definition

makeclean committed:
updated signatures according to new changes in how seed is set

makeclean committed:
updated signatures according to new changes in how seed is set

eepeterson committed:
fixed general cylinder _to_base_coeffs method

makeclean committed:
Updated the source routines according to passing seed to prn

makeclean committed:
Updated the source routines according to passing seed to prn

makeclean committed:
Updated documentation to reflect changes

makeclean committed:
Updated documentation to reflect changes

makeclean committed:
Added statements for additional cmake targets

makeclean committed:
Added statements for additional cmake targets

makeclean committed:
Updated spelling mistakes

makeclean committed:
Updated spelling mistakes

makeclean committed:
Corrected style according to review

makeclean committed:
Corrected style according to review

makeclean committed:
line removal from review

makeclean committed:
line removal from review

makeclean committed:
further changes from review

makeclean committed:
further changes from review

makeclean committed:
further review comments regarding style

makeclean committed:
further review comments regarding style

makeclean committed:
Refactor into function as suggested by review

makeclean committed:
Refactor into function as suggested by review

makeclean committed:
updated for new cmake instructions

makeclean committed:
updated for new cmake instructions

makeclean committed:
updated the example source according to style

makeclean committed:
updated the example source according to style

eepeterson committed:
implemented helper functions

eepeterson committed:
finished general cone transformation

eepeterson committed:
finished implementation

eepeterson committed:
implemented individual translate and evaulate methods and removed extraneous docstrings

eepeterson committed:
fixed pytest results after ci failure

eepeterson committed:
applied some suggestions from code review

eepeterson committed:
refactored translate method back out to mixins

paulromano committed:
Fixes to custom source feature

Mikolaj-A-Kowalski committed:
Enrichment gives exceptions for non 2 isotopes

Following discussion in the issues the functionality to enrich mixtures
of more then 2 nuclides is block by an exception. The algorithim should
still support larger mixtures.

eepeterson committed:
implemented Plane.__new__ for returning simplified versions

Mikolaj-A-Kowalski committed:
Implemented unit tests for Element.expand

Mikolaj-A-Kowalski committed:
Adds general element enrichment to Material

Adds an option to specify an enrichment of an element composed of two
isotopes to Material.add_element

Mikolaj-A-Kowalski committed:
Merge remote-tracking branch 'upstream/develop' into cam_WP7_task3

paulromano committed:
Avoid copysign in instances where an int return value is needed

makeclean committed:
Merge branch 'dlopen_source' of https://github.com/makeclean/openmc into dlopen_source

makeclean committed:
revert previous attempts to fix

makeclean committed:
Merge pull request #2 from paulromano/dlopen_source

Fixes to custom source feature

makeclean committed:
updated test signature

eepeterson committed:
reverted simplifying Plane constructor and default args

eepeterson committed:
removed debug statement

makeclean committed:
updated test to use new build method

SamPUG committed:
Fixed bug which prevents creating photon data with no atomic relax

SamPUG committed:
Unit test for exporting photon data with no atomic relax

eepeterson committed:
Apply suggestions from code review

Co-Authored-By: Paul Romano <paul.k.romano@gmail.com>

eepeterson committed:
applied more suggestions from code review and added coefficient cache for general cones and cylinders

eepeterson committed:
merged remote into local after review

paulromano committed:
Get the source_dlopen test working

eepeterson committed:
removed coefficient caching

paulromano committed:
Merge pull request #1483 from eepeterson/refactor_surface

Refactor surface.py code for easier extension

amandalund committed:
Merge pull request #1486 from paulromano/copysign-fix

Fix bug occurring when surface ID > 999999

SamPUG committed:
Added suggested changes for #1489

SamPUG committed:
Made requested formatting change

paulromano committed:
Merge pull request #1489 from SamPUG/develop

Fixed bug which prevents creating photon data with no atomic relax

paulromano committed:
Make sure DAGMC is discoverable if linking against OpenMC

eepeterson committed:
Refactoring improvements

eepeterson committed:
removed __init_subclass__ since it's not in python 3.5

eepeterson committed:
removed extraneous definition

paulromano committed:
Don't use target_link_directorires (only supported in cmake 3.13+)

simondrichards commented on tests/unit_tests/test_element.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def test_expand_no_ernichment():
+    """ Expand Li in natural compositions"""
```
There is a typo in the test name: should be corrected to test_expand_no_e**nr**ichment

simondrichards commented on tests/unit_tests/test_element.py:
```diff
@@ -0,0 +1,75 @@
+    for isotope in lithium.expand(100.0, 'ao'):
+        assert isotope[1] == pt.approx(NATURAL_ABUNDANCE[isotope[0]] * 100.0, rel=TOL)
+
+    # Verify the expanction into WEIGHT fraction against natural composition
```
Minor typo in comment - "expanction" should be "expansion"

simondrichards commented on tests/unit_tests/test_element.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def test_expand_exceptions():
+    """ Test that correct exceptions are raiused for invalid input """
```
Minor typo in docstring - "raiused"

simondrichards commented on openmc/element.py:
```diff
@@ -65,16 +74,55 @@ def expand(self, percent, percent_type, enrichment=None,
+
+        When the `enrichment` argument is specified with `enrichment_target` a
+        general enrichment procedure is used. It will raise exception unless
+        element is composed of excatly 2 isotopes. `enrichment` is interpreted
```
Minor typo in comments : "excatly"

simondrichards commented on openmc/element.py:
```diff
@@ -181,6 +236,70 @@ def expand(self, percent, percent_type, enrichment=None,
+        # Interpret required enrichment as weight
+        elif enrichment is not None and enrichment_target is not None:
+
+            # Check if is a single isotope mixture
```
Is the comment correct here? Should this say "# Check if it is a two-isotope mixture"?

simondrichards commented on openmc/element.py:
```diff
@@ -181,6 +236,70 @@ def expand(self, percent, percent_type, enrichment=None,
+            if len(abundances) != 2:
+                msg = 'Element {0} does not consist of 2 isotopes. Thus it '\
+                      'cannot be enriched with in-build procedure. Please '\
+                      'enter isotopic abundances manually. '.format(self)
```
Suggest minor rewording:
`             msg = 'Element {0} does not consist of 2 naturally-occurring isotopes. '\
                           'Therefore it cannot be enriched with the in-built procedure. Please '\
                       'enter isotopic abundances manually. '.format(self)`

simondrichards commented on openmc/element.py:
```diff
@@ -164,7 +212,14 @@ def expand(self, percent, percent_type, enrichment=None,
         # Modify mole fractions if enrichment provided
-        if enrichment is not None:
+        # Old treatment for Uranium
+        if enrichment is not None and enrichment_target is None:
```
I wonder if this condition should be extended to include the case where enrichment_target=='U235'?

Currently if anyone tried using this function for uranium and explicitly set the enrichment_target to U235 it would raise an error, because uranium has more than 2 natural isotopes. The error message will say that uranium can't be enriched with the inbuilt procedure, but of course that is not correct.

simondrichards commented on openmc/material.py:
```diff
@@ -512,9 +513,24 @@ def add_element(self, element, percent, percent_type='ao', enrichment=None):
+        Notes
+        -----
+        General enrichment procedure is allowed only for elements composed of
+        two isotopes. If `enrichment_target` is given withou `enrichment`
```
withou --> without

simondrichards reviewed (commented):
This mostly looks good to me. Just a few minor comments pointing out typos. As noted in the comments, we could allow the case where uranium is enriched with the specific enrichment target of U235. Currently this is trapped but with a misleading error message.

simondrichards commented:
In addition to the comments on the code and tests, I guess you should also update the user guide section in docs/source/usersguide/materials.rst to describe the added functionality for enriching elements other than uranium.

eepeterson committed:
Apply suggestions from code review

Co-Authored-By: Paul Romano <paul.k.romano@gmail.com>

eepeterson committed:
applied suggestions from code review

pshriwise committed:
Merge pull request #1490 from paulromano/dagmc-target

Make sure DAGMC is discoverable if linking against OpenMC

eepeterson committed:
removed use of round in favor of isclose

eepeterson committed:
finished unit tests and fixed wrong math

makeclean committed:
Merge branch 'develop' into dlopen_source

makeclean committed:
Merge pull request #3 from paulromano/dlopen_source

Get the source_dlopen test working

paulromano committed:
Revert changes in CMakeLists.txt

paulromano committed:
Documentation updates

makeclean committed:
Merge pull request #4 from paulromano/dlopen_source

A few more updates to your branch

eepeterson committed:
applied suggestions from code review

Commit:
Disallow numpy ufuncs being applied directly to FissionYields

Set ``__array_ufunc__`` to ``None`` to ensure that numpy floats respect
__rmul__ and __radd__ for fission yields. Allows weights to be left
or right multiplied as numpy floats, which can be computed during
fission yield weighting. A unit test has been added to ensure this
behavior is preserved

Closes #1492

paulromano committed:
Merge pull request #1257 from makeclean/dlopen_source

Dynamically loaded custom source function

paulromano committed:
Merge pull request #1491 from eepeterson/surface_refactoring2

Surface refactoring improvements

paulromano committed:
Merge pull request #1493 from drewejohnson/numpy-mul-fy

Disallow numpy ufuncs being applied directly to FissionYields

Mikolaj-A-Kowalski committed:
Merge branch 'develop' into cam_WP7_task3

Mikolaj-A-Kowalski committed:
Address comments by @simondrichards

Also remove spaces between Raises entries in docstring of Element.expand

Mikolaj-A-Kowalski commented:
Hopefully this latest commit should address all the comments.  
I have corrected the typos. 
Provided more informative error message for U235 target 
Updated Material section of the User Guide

simondrichards commented on docs/source/usersguide/materials.rst:
```diff
@@ -55,6 +55,20 @@ following would add 3.2% enriched uranium to a material::
 
+It is also possible to perform enrichment of any element that is composed
+of two naturally-occurring isotopes (e.g. Li, B) in terms of atomic percent.
+To invoke it, provide additional arguments `enrichment_target` to
```
I suggest a slight rewording of this line:
"To invoke this, provide the additional argument `enrichment_target` to".

simondrichards commented on docs/source/usersguide/materials.rst:
```diff
@@ -55,6 +55,20 @@ following would add 3.2% enriched uranium to a material::
+
+   mat.add_element('B', 1.0, enrichment=30.0, enrichment_target='B10')
+
+In order to enrich an isotope in terms of mass percent (wo%). Provide extra
```
I suggest a minor change to this line:
"In order to enrich an isotope in terms of mass percent (wo%), provide the extra"

simondrichards commented on openmc/element.py:
```diff
@@ -241,11 +235,21 @@ def expand(self, percent, percent_type, enrichment=None,
+            if enrichment_target == 'U235':
+                msg = "There is a special procedure for enrichment of U235 "\
+                      "in U. To invoke it do not specify neither "\
+                      "'enrichment_target' nor 'enrichment_type'. Provide "\
```
I suggest a slight rewording of the second sentence:
"To invoke it, the arguments 'enrichment_target' and 'enrichment_type' should be omitted."

simondrichards commented:
Thanks for the updates @Mikolaj-A-Kowalski. I've suggested some minor rewording and then I think this pull request is ready to merge.

Mikolaj-A-Kowalski committed:
Address further comments by @simondrichards

Corrects spelling and grammar.

simondrichards commented:
Thanks @Mikolaj-A-Kowalski - pull request merged and ready to submit a pull request to openmc-dev.
